tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -0,0 +1,709 @@
+            if not 'no_mac' in bazel_tags:
+                platforms.append('mac')
+            if not 'no_windows' in bazel_tags:
+                platforms.append('windows')
+            test_dict['platforms'] = platforms
+
+        if '//external:benchmark' in bazel_rule['transitive_deps']:
+            test_dict['benchmark'] = True
+            test_dict['defaults'] = 'benchmark'
+
+        cmdline_args = bazel_rule['args']
+        if cmdline_args:
+            test_dict['args'] = list(cmdline_args)
+
+        uses_gtest = '//external:gtest' in bazel_rule['transitive_deps']
+        if uses_gtest:
+            test_dict['gtest'] = True
+
+        if test.startswith('test/cpp') or uses_gtest:
+            test_dict['language'] = 'c++'
+            
+        elif test.startswith('test/core'):
+            test_dict['language'] = 'c'
+        else:
+            raise Exception('wrong test' + test)
+
+        # short test name without the path.
+        # There can be name collisions, but we will resolve them later
+        simple_test_name = os.path.basename(_extract_source_file_path(test))
+        test_dict['_RENAME'] = simple_test_name
+
+        test_metadata[test] = test_dict
+
+    # detect duplicate test names
+    tests_by_simple_name = {}
+    for test_name, test_dict in test_metadata.iteritems():
+        simple_test_name = test_dict['_RENAME']
+        if not simple_test_name in tests_by_simple_name:
+            tests_by_simple_name[simple_test_name] = []
+        tests_by_simple_name[simple_test_name].append(test_name)
+
+    # choose alternative names for tests with a name collision
+    for collision_list in tests_by_simple_name.itervalues():
+        if len(collision_list) > 1:
+            for test_name in collision_list:
+                long_name = test_name.replace('/', '_').replace(':', '_')
+                print('short name of "%s" collides with another test, renaming to %s' % (test_name, long_name))
+                test_metadata[test_name]['_RENAME'] = long_name
+
+    # TODO(jtattermusch): in bazel, add "_test" suffix to the test names
+    # test does not have "_test" suffix: fling
+    # test does not have "_test" suffix: fling_stream
+    # test does not have "_test" suffix: client_ssl
+    # test does not have "_test" suffix: handshake_server_with_readahead_handshaker
+    # test does not have "_test" suffix: handshake_verify_peer_options
+    # test does not have "_test" suffix: server_ssl
+
+    return test_metadata
+
+# extra metadata that will be used to construct build.yaml
+# there are mostly extra properties that we weren't able to obtain from the bazel build
+# _TYPE: whether this is library, target or test
+# _RENAME: whether this target should be renamed to a different name (to match expectations of make and cmake builds)
+# NOTE: secure is 'check' by default, so setting secure = False below does matter
+_BUILD_EXTRA_METADATA = {
```

Comment:
How about having a separate file like `build_metadata.yaml` to keep this script isolated?

Reply:
Makes sense, if you're fine with it, I'd leave this as a followup work (I don't want to further complicate this PR). Once merged, changed this will be easy (and we will have a better feel what are the best nest steps).